presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushPartialAggregationThroughJoin.java

```diff
@@ -0,0 +1,186 @@
+    }
+
+    private PlanNode pushPartialToRightChild(AggregationNode node, JoinNode child)
+    {
+        List<Symbol> groupingSet = getPushedDownGroupingSet(node, child, ImmutableSet.copyOf(child.getRight().getOutputSymbols()));
+        AggregationNode pushedAggregation = replaceAggregationSource(node, child.getRight(), child.getCriteria(), groupingSet);
+        return pushPartialToJoin(pushedAggregation, child, child.getLeft(), pushedAggregation, child.getLeft().getOutputSymbols());
+    }
+
+    private List<Symbol> getPushedDownGroupingSet(AggregationNode aggregation, JoinNode join, Set<Symbol> availableSymbols)
+    {
+        List<Symbol> groupingSet = Iterables.getOnlyElement(aggregation.getGroupingSets());
+        Set<Symbol> joinKeys = Stream.concat(
+                join.getCriteria().stream().map(JoinNode.EquiJoinClause::getLeft),
+                join.getCriteria().stream().map(JoinNode.EquiJoinClause::getRight)
+        ).collect(Collectors.toSet());
+
+        // keep symbols that are either directly from the join's child (availableSymbols) or there is
+        // an equality in join condition to a symbol for the join child
+        List<Symbol> pushedDownGroupingSet = groupingSet.stream()
+                .filter(symbol -> joinKeys.contains(symbol) || availableSymbols.contains(symbol))
+                .collect(Collectors.toList());
+
+        if (pushedDownGroupingSet.size() != groupingSet.size() || pushedDownGroupingSet.isEmpty()) {
+            // If we dropped some symbol, we have to add all join key columns to the grouping set
+            Set<Symbol> existingSymbols = ImmutableSet.copyOf(pushedDownGroupingSet);
+
+            join.getCriteria().stream()
+                    .filter(equiJoinClause -> !existingSymbols.contains(equiJoinClause.getLeft()) && !existingSymbols.contains(equiJoinClause.getRight()))
+                    .forEach(joinClause -> pushedDownGroupingSet.add(joinClause.getLeft()));
+        }
+        return pushedDownGroupingSet;
+    }
+
+    private AggregationNode replaceAggregationSource(
+            AggregationNode aggregation,
+            PlanNode source,
+            List<JoinNode.EquiJoinClause> criteria,
+            List<Symbol> groupingSet)
+    {
+        ImmutableSet<Symbol> sourceSymbols = ImmutableSet.copyOf(source.getOutputSymbols());
+        ImmutableMap.Builder<Symbol, Symbol> mapping = ImmutableMap.builder();
+
+        for (JoinNode.EquiJoinClause joinClause : criteria) {
+            if (sourceSymbols.contains(joinClause.getLeft())) {
+                mapping.put(joinClause.getRight(), joinClause.getLeft());
+            }
+            else {
+                mapping.put(joinClause.getLeft(), joinClause.getRight());
+            }
+        }
+
+        AggregationNode pushedAggregation = new AggregationNode(
+                aggregation.getId(),
+                aggregation.getSource(),
```

Comment:
I thought that source of aggregation is join, so you should use `source` here iMO